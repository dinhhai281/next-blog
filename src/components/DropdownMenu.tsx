import { ChevronDownIcon } from '@chakra-ui/icons';
import {
  Box,
  Button,
  ButtonProps,
  Popover,
  PopoverBody,
  PopoverContent,
  PopoverTrigger,
  Text,
  useDisclosure,
} from '@chakra-ui/react';
import React, { FC, ReactNode } from 'react';

import { MotionBox } from './MotionBox';

export interface DropdownMenuProps extends ButtonProps {
  items?: any[];
  label: string;
  children: (props: any) => ReactNode;
}

export const DropdownMenu: FC<DropdownMenuProps> = ({ children, label, items = [], ...rest }) => {
  const { isOpen, onOpen, onClose } = useDisclosure();

  return (
    <Popover trigger='hover' onOpen={onOpen} isOpen={isOpen} onClose={onClose}>
      <PopoverTrigger>
        <Button
          variant='ghost'
          iconSpacing={1}
          as='div'
          cursor='pointer'
          {...rest}
          rightIcon={
            <MotionBox
              animate={{
                rotate: isOpen ? -180 : 0,
              }}
            >
              <ChevronDownIcon boxSize={5} />
            </MotionBox>
          }
        >
          <Text fontSize='lg' fontWeight={400}>
            {label}
          </Text>
        </Button>
      </PopoverTrigger>
      <PopoverContent>
        <PopoverBody>
          {items?.map(item => (
            <Box key={item} onClick={onClose}>
              {children ? (
                children(item)
              ) : (
                <Button variant='ghost' isFullWidth justifyContent='flex-start' h={12}>
                  {item}
                </Button>
              )}
            </Box>
          ))}
        </PopoverBody>
      </PopoverContent>
    </Popover>
  );
};
