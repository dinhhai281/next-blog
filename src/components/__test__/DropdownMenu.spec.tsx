import { render } from '@testing-library/react';
import React from 'react';

import { DropdownMenu } from '../DropdownMenu';

describe('DropdownMenu', () => {
  const renderComponent = () =>
    render(
      <DropdownMenu>
        <span>title</span>
      </DropdownMenu>
    );
  it('should match snapshot', () => {
    const { asFragment } = renderComponent();
    expect(asFragment()).toMatchSnapshot();
  });
});
