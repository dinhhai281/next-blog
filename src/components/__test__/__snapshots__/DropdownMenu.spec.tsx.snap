// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu should match snapshot 1`] = `
<DocumentFragment>
  <div
    aria-controls="popover-content-2"
    aria-expanded="false"
    aria-haspopup="dialog"
    class="chakra-button css-jt8z5c"
    id="popover-trigger-2"
  >
    <p
      class="chakra-text css-191iiyz"
    />
    <span
      class="chakra-button__icon css-6eepfv"
    >
      <div
        aria-hidden="true"
        class="css-0"
        focusable="false"
      >
        <svg
          class="chakra-icon css-w31a5m"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            fill="currentColor"
          />
        </svg>
      </div>
    </span>
  </div>
  <div
    class="css-0"
    style="visibility: hidden; position: absolute; left: 0px; top: 0px;"
  >
    <section
      aria-describedby="popover-body-2"
      class="chakra-popover__content css-1ab1jlb"
      id="popover-content-2"
      role="tooltip"
      style="transform-origin: top center; opacity: 0; visibility: hidden; transform: scale(0.95) translateZ(0);"
      tabindex="-1"
    >
      <div
        class="chakra-popover__body css-0"
        id="popover-body-2"
      />
    </section>
  </div>
</DocumentFragment>
`;
